### ISSUE TITLE
no examples on C#7 support in self host environment that was added last release

### ISSUE THREAD

**Issue by seertenedos** [oleg-shilo/cs-script#74]
In the nuget notes it states that c#7 support was added but so far i can't work out how to get the c#7 syntax to function with the CodeDom compiler.  There is some doco for when you are using the standalone app but we are not.  The scripts are embedded in our application and we still need the debug support.  Main syntax we are trying to use is the xxx?.yyy for handling nulls.

If i change the evaluator to Rosolyn it works but then i can't debug and i can't find the nuget package that contains everything needed to make c#7 work with CodeDom.

**Comment by oleg-shilo** [comment]
I will update the doco. Txs.

This is how you can do this:
```C#
CSScript.GlobalSettings.UseAlternativeCompiler = @"E:\cs-script\lib\CSSRoslynProvider.dll";
var sayHello = CSScript.LoadMethod(@"static void SayHello(string greeting)
                                    {
                                        var tuple = (1,2);
                                        void test()
                                        {
                                            Console.WriteLine(""Hello from C#7"");
                                            Console.WriteLine(tuple);
                                        }
                                        test();

                                        Console.WriteLine(greeting);
                                    }")
                        .GetStaticMethod();

sayHello("Hello World!");
```

![image](https://user-images.githubusercontent.com/16729806/27633818-192f7630-5c44-11e7-9640-e51f20d447a9.png)

**Comment by seertenedos** [comment]
Thanks for the info. Where can I get CSSRoslynProvider.dll as I pulled all
the ca-script nuget's and I could not see it in any of them or am I missing
something? I am hoping I can do this via nuget as our scripting support
sits in one of our own nuget librarys that just includes a few base classes
and helpers our scripts need.

**Comment by oleg-shilo** [comment]
I will be adding it as a separate NuGet package (it is still under development). Choco is still not released. Matter of a few days.

For now you can get is with v3.26.0.0 distro from GitHub (https://github.com/oleg-shilo/cs-script/releases).

Just have the whole Lib directory stored locally and indicate the location of the CSSRoslynCompiler.dll as In the code sample.

**Comment by seertenedos** [comment]
thanks. does that mean that in future it could be done without specifying a
path if you have a reference to the dll or something? Anyway i will look at
it tomorrow when i get into the office but that should get me running for
now.

**Comment by seertenedos** [comment]
thanks i can confirmed that is working for me though for now it will only be a solution that works on my pc till the nuget package is done so it will all come packaged in the project and work on other dev pcs, build servers and destination servers.

**Comment by oleg-shilo** [comment]
Now you can add to your project an extra NuGet package (comes with the code sample):
```
Install-Package CS-Script.RoslynProvider
```

**Comment by seertenedos** [comment]
Thanks i will play around with it today. Any chance of doing a second version that does copy the files to the output folder? At the moment i am trying to work out how to make my nuspec references yours as well as copy the required files over. I am using on dotnet 4.6 so without copying the files it will not work when deployed. I get it that dotnet core is different as it will just restore the files for you.

**Comment by oleg-shilo** [comment]
The sample included in the package shows the technique.

**Comment by seertenedos** [comment]
That only works when the packages folder exists which will only be there in
dotnet core released apps. Normal deployed dotnet apps don't have a
packages folder once they are built and deployed.

**Comment by seertenedos** [comment]
also even with the dll itself i found that when you have another project
that references a project referencing and using the scripting library (even
a unit test project) then that project needs to directly have the nuget
package added otherwise it can't find or use the RoslynProvider.dll as it
only gets added to the first projects output folder where as CS-Script.bin
did not need to be directly referenced in all projects that reference a
library using it.

**Comment by seertenedos** [comment]
maybe making it an embedded resource or something may work better for that
dll and that would give you a fixed location to reference it add i guess.
Still trying to find an option that works for me.

**Comment by oleg-shilo** [comment]
Roslyn deployment model is plain ridiculous. 
- Roslyn is not part of the .NET distro. You have to deploy it manually. Guess what... MONO includes Roslyn. So you don't need to distribute 16M with every application that uses Roslyn.
- On .NET environment it has to be deployed in the hard-coded  `./bin/roslyn` with respect to the host app. 
- Even if Roslyn binaries are present it cannot be used with the default `CSharpCodeProvider` and an extra NuGet package has to be used.
- Even when that extra package (Microsoft.CodeDom.Providers.DotNetCompilerPlatform) is used it cannot be configured to use Roslyn compiler "cscs.exe" from any other location but in the hard-codded one. _ASP.NET team refuses any bug reports insisting you never need to have compilers in any other location but  `./bin/roslyn`._
- Even if you solved this problem Microsoft.CodeDom.Providers.DotNetCompilerPlatform still cannot be used under Mono so forget Linux.
- Authors of Roslyn compilers NuGet package decided that you never need to copy compilers in your  "output" directory as you mentioned. This is in a complete contrast to the Microsoft.CodeDom.Providers.DotNetCompilerPlatform authors that insist on the binaries being only in "output" folder. 
- There is another NuGet package (Microsoft.CodeDom.Providers.DotNetCompilerPlatform.BinFix) which only designed for address the problem of copying all roslyn files in the output folder. But... it didn't give me any joy.

CS-Script cannot and should not fix all this mess. 

Though CS-Script made it possible to have a work around. CS-Script includes [CSSRoslynProvider](https://github.com/oleg-shilo/cs-script/tree/master/Source/CSSRoslynProvider) which embeds patched version of Microsoft.CodeDom.Providers.DotNetCompilerPlatform. CSSRoalynProvider makes it possible to use Roslyn from any location. This can be global as on Mono/Linux or any other as on .NET/Win. 

The approach CSSRoslynProvider uses is simple and it's consistent with Mono approach. Have your Roslyn binaries anywhere you want. Use NuGet or any other deployment model. Just set it's location at runtime.
```C#
CSScript.GlobalSettings.UseAlternativeCompiler = <CSSRoslynProvider.dll path>;
CSScript.GlobalSettings.RoslynDir = <Roslyn path>;
``` 
Though if the binaries are found in the default locations then you can skip the assignment above.

But of course deployment of Roslyn is completely responsibility of the application vendor. This unfortunately cannot be changed.

**Comment by oleg-shilo** [comment]
> hen that project needs to directly have the nuget
package added otherwise it can't find or use the RoslynProvider.dll as it
only gets added to the first projects output folder where as CS-Script.bin

Sorry but this is how NuGet works. I have no control over it. You need to add the package or copy the required assemblies in the pre-build event.

**Comment by seertenedos** [comment]
is it because nothing references the dll?  You can do embedded etc if you
use "contentFiles" in the nuget spec. would it be possible to
make CSScript.GlobalSettings.UseAlternativeCompiler accept an embedded file
path or have another method that takes an embedded file path or is there
some other way to make this work? I think i have the other Roslyn issues
fixed somewhat with copy to output folder but still need to do mroe testing
there. Otherwise i will embed them as well and extract as part of starting
up though that is not ideal.

**Comment by oleg-shilo** [comment]
> maybe making it an embedded resource or something may work better for that
dll and that would give you a fixed location to reference it add i guess.

Yes this is the solution I am leaning towards to. I have started preparations for that by embedding "Microsoft.CodeDom.Providers.DotNetCompilerPlatform" into "CSSRoalynProvider". So you already have one dll less. 

Most likely in the next releases I will embed CSSRoalynProvider into cscs.exe and CSScriptLibrary.dll but I am reluctant to promise that as one may as "What about Mono.CSharp.dll? Should it be also embedded?". And you :) may also ask "What about embedding Roslyn?"

You see how challenging it is?

**Comment by seertenedos** [comment]
yea i can but then that is what different nugets are for. We use your
scripting all though about 6 applications and 9 different libraries
excluding unit tests but it all feeds from one project that wraps your
library to give us caching of scripts etc. I need to get this to the level
where i can change our central nuget package and the rest should work as it
flows up an down the line.  That was easy before but with Roslyn it has
become impossible at the minute.  Even using ncrunch for unit tests means i
need custom locator code for the files for when it uses vs MSTest vs
running on build server vs running on visual studio project reference vs
nuget reference vs deployment. Currently it is a real mess but i need to
get c#7 support so i will keep plugging away at it even if i need to build
a custom nuget that just has the references in it and embeds them in a dll
that just has a single class that extracts the files to fixed location and
auto configures CS-Script to use it.

**Comment by oleg-shilo** [comment]
> yea i can but then that is what different nugets are for...

Agree. But... if you are happy to add 16M Roslyn NuGet package to every project then why you cannot do the same for 2M CSSRoslynProvider+CS-Script?

Anyway, I do believe that embedding CSSRoslynProvider (into CSSCriptLibrary) will solve the problem of carrying it through your project outputs but you will have to wait for the next release. BTW please log it as an enhancement request.

I think the most reasonable solution for you right now is either to reference CSSRoslynProvider.dll or to have it as a non compileable project content that is copied into the output folder. 

As for Roslyn binaries I would have the location set as it "CSSCRIPT_ROSLYN" global or process environment variable. It is an alternative to `CSScript.GlobalSettings.RoslynDir = <Roslyn path>;`

**Comment by seertenedos** [comment]
CSSRoslynProvider is referenced but is treated differently for some
reason.  I don't plan to reference a 16M nuget package everywhere and it is
not the size that bothers me it is that i need to change every project that
uses it even indirectly.  I plan to do the embedding in a dll that is
already referenced and already flows though our system without every
library needing to add the nuget if they reference a library that uses it
etc. I will also raise the enhancement in a minute

**Comment by seertenedos** [comment]
Enhancement: Embedded Roslyn support in separate nuget #76

**Comment by oleg-shilo** [comment]
&gt; _CSSRoslynProvider is referenced but is treated differently for some
reason._

I just checked with a fresh _console app_ project and it references and copies _CSSRoslynProvider.dll_  into output directory correctly. The only one thing to be aware is that the package references the older version of CS-Script so you need to update it from the NuGet  Package manager:
![image](https://user-images.githubusercontent.com/16729806/27792247-207f45a4-603c-11e7-92e2-ea46c7fccb2c.png)

----

&gt; [_From #76] ...The required files should also be embedded and either accessed as an embedded resource or extracted locally when the methods on the static classes to configure CS-Script to use Roslyn is called._

I am afraid I gave you a wrong impression when I asked you to log a feature request.  I was talking about "embedding CSSRoslynProvider (into CSSCriptLibrary) " but not about embedding Roslyn itself.

 The diagram below helps to understand the issue:
![image](https://user-images.githubusercontent.com/16729806/27790724-2545a24c-6035-11e7-8e5b-26807653b3a0.png)

The `Microsoft.CodeDom.Providers.DotNetCompilerPlatform.dll` is already embedded into `CSSRoslynProvider.dll` thus it is only a singe dll. 
I was also considering as a next step to embed `CSSRoslynProvider.dll` into `cscs.exe` and `CSSCriptLibrary.dll` and this is what I had in mind for that feature request. 

However embedding the whole Roslyn is conceptually questionable. And what is even more important - technically impossible. Roslyn binaries (as opposite to CSSRoslynProvider.dll) cannot be consumed from the resources. They need to be present on the file system. However "the presence" can be already  achieved with "Microsoft.Net.Compilers" package. The remaining part of the Roslyn integration "setting CSScript.GlobalSettings.RoslynDir" can be done easily from the assembly initialization routine of the test project. Thus the whole objective hardly justifies the development of a special package.

Of course if you strongly feel about this or I misinterpret your idea you can always develop that small highly specialized package and I will be happy to endorse it in the CS-Script documentation.

**Comment by seertenedos** [comment]
This is driving me nuts and not your fault at all.  In our case everything is added to a library project that is packed into a nuget. This nuget is then referenced in a number of our libraries in our actual projects that is then project referenced to the console/windows services.  I have managed to sort out the issues i was having with the CSSRoslynProvider.dll by supplying 4 different search paths which then covers all the different paths it ends up in when referenced in the different ways and in the different use cases. (mstest vs xunit vs ncrunch caused a lot of hassle here) with our whole build process that only took 4 hours to sort out. I have now spent about 6 hours on trying to get the Roslyn compilers to be able to be located but it still does not work in all cases.  Mainly this one fails on the final deployment. I have been playing with the "contentFiles" part of the nuget spec and the "copyToOutput" setting but for the life of me i can't get this to work.

There is no reason why an embedded copy of the files should not work it would just be tricky to do at least in our build environments. Yes they need to be on the disk to run but that can be the first step on the class you call to set it up. If the required files are not in fixed relative location then extract them from embedded resource and save them to the disk. If i can't get my nuget to work soon i will have to manually go down this path.

**Comment by oleg-shilo** [comment]
Yes the "mstest vs xunit vs ncrunch" bunch is not playing nice with the dependencies. Particularly mstest (at least from my experience). 

Roslyn makes it even worse. I cannot believe MS is still insisting that Roslyn compilers are not system wide resource but a project one. It should never be a part of any NuGet package but a system setup (msi or choco). It should be installed once and easily available to all interested parties. Like it's done on Linux/Mono.

Anyway, a potential work around can be to make Roslyn a part of your custom NuGet package and upon installation (probably from ps1 script) set `CSSCRIPT_ROSLYN` system env var to the location of the just installed Roslyn. This would completely eliminate the necessity for copying it around and even `CSScript.GlobalSettings.RoslynDir = <Roslyn path>` will be no longer needed. That envvar effectively makes Roslyn a system wide resource.

**Comment by seertenedos** [comment]
Tried a lot of that sort of thing but vs2017 does not support ps1 script
from what I understand and the build server was still causing me issues.
The solution was simple in the end. Took a copy of the compiler files and
embedded it in our scripting dll. On start up it checks all the embedded
files exist on disk and extracts any that are missing to the set folder.

Works in every use case over all systems I tested. I did not do the same
for your dll so still had a few small issues there but system is working
now.
